File: go/wb_meta.go
// This file is part of dm-writeboost-tools
// Copyright (C) 2015 Akira Hayakawa <ruby.wktk@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

package main

import (
	"encoding/binary"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strconv"

	"./lib"
)

type sb_header struct {
	magic uint32
}

type sb_record struct {
}

type metablock struct {
	sector     uint64
	dirty_bits uint8
}

// TODO help

func read_metablock(f io.Reader) metablock {
	var v metablock
	binary.Read(f, binary.LittleEndian, &v.sector)
	binary.Read(f, binary.LittleEndian, &v.dirty_bits)
	io.CopyN(ioutil.Discard, f, 16-(8+1))
	return v
}

func main() {
	dev := lib.NewDev(os.Args[1])
	// fmt.Println(dev.Size())
	id, _ := strconv.Atoi(os.Args[2])
	if id == 0 { // superblock
		f, _ := os.Open(dev.Name())
		var v32 uint32
		binary.Read(f, binary.LittleEndian, &v32)
		var s string
		if v32 == 0x57427374 {
			s = "formatted"
		} else {
			s = "not formatted"
		}
		fmt.Println("superblock header")
		fmt.Printf("magic number: 0x%x (%s)\n", v32, s)

		fmt.Println()

		var v64 uint64
		f.Seek((1<<20)-512, 0)
		binary.Read(f, binary.LittleEndian, &v64)
		fmt.Println("superblock record")
		fmt.Println("last writeback id:", v64)
	} else { // segment header (id > 0)
		f, _ := os.Open(dev.Name())
		f.Seek(int64(dev.CalcSegmentStart(id))<<9, 0)
		segH := lib.ReadSegmentHeader(f)
		fmt.Println("id:", segH.Id)
		fmt.Println("checksum:", segH.Checksum)
		fmt.Println("length:", segH.Length)
		fmt.Println()
		for i := 0; i < 127; i++ {
			mb := read_metablock(f)
			fmt.Printf("%3d sector=%d dirty_bits=%d\n", i, mb.sector, mb.dirty_bits)
		}
	}
}

File: src/wbmeta.rs
extern crate clap;

use std::fs::File;
use std::io::Read;
use std::io::Seek;
use std::io::SeekFrom;

use clap::Parser;
#[derive(Parser)]
#[command(name = "wbmeta")]
#[command(about = "Dump a segment header")]
#[command(author, version)]
struct Args {
    #[arg(help = "Path to the cache device")]
    cachedev: String,
    #[arg(help = "Segment id. 0 means the superblock")]
    segid: i32,
}

fn main() {
    let args = Args::parse();

    let devname: String = args.cachedev;
    let id: i32 = args.segid;

    let cache_dev = lib::CacheDevice::new(devname.to_owned());

    if id == 0 {
        // superblock
        let mut buf = [0; 512];
        let mut f = File::open(&devname).unwrap();
        f.read(&mut buf).unwrap();
        let sup_header = lib::SuperBlockHeader::from_buf(&buf);

        let s = if sup_header.magic == 0x57427374 {
            "formatted"
        } else {
            "unformatted"
        };

        println!("[superblock header]");
        println!("magic = {} ({})", sup_header.magic, s);

        f.seek(SeekFrom::Start((1u64 << 20) - 512)).unwrap();
        f.read(&mut buf).unwrap();
        let sup_record = lib::SuperBlockRecord::from_buf(&buf);

        println!("[superblock record]");
        println!(
            "last writeback id = {}",
            sup_record.last_writeback_segment_id
        );
    } else {
        // header
        let mut buf = [0; 4096];
        let mut f = File::open(&devname).unwrap();
        let start_byte: u64 = (cache_dev.calc_segment_start(id) as u64) << 9;
        f.seek(SeekFrom::Start(start_byte)).unwrap();
        f.read(&mut buf).unwrap();
        let (header, metablocks) = lib::Segment::from_buf(&buf);

        println!("[segment header]");
        println!("id        = {}", header.id);
        println!("checksumx = {}", header.checksum);
        println!("length    = {}", header.length);

        for (i, metablock) in metablocks.iter().enumerate() {
            println!(
                "[{}] sector={}, dirty_bits={}",
                i, metablock.sector, metablock.dirty_bits
            );
        }
    }
}
